### DISCUSSION TITLE
Differences between semantic-kernel and langchain?

### DISCUSSION THREAD

**Discussion by SeaDude** [microsoft/semantic-kernel#1326] [Q&A]
Hello,

Seems like these two projects have a lot in common. What are the definite differences?

Thank you

**Comment by hathind-ms** [comment]
Hi @SeaDude,

We haven't done a detailed comparison. This space is moving very quickly, and our objective has been to move as fast as possible reacting to feedback and feature asks from devs working in any language.

Thanks!

**Comment by santanaeds** [comment]
As far as I know, semantic-kernel do support .Net, which langchain doesn't, that's all.

**Reply by tonyqus** [reply]
But .NET is not the leading langauge in machine learning or AI. Does that means SK is only for developers who don't wanna learn Python?

**Comment by adrianwyatt** [comment]
Overheard from the /Build conference: LangChain is an open-source alternative to Semantic Kernel for AI orchestration alongside other options like Haystack, Hugging Face Agents Transformers. Each AI orchestrator has different strengths and weaknesses. Semantic Kernel's strength is how it interlocks with the many existing and emergent AI Platform components at Microsoft while still enabling completely OPEN access to components outside of the Microsoft stack by design.

**Comment by priya-27** [comment]
If we talk in terms of python, what are the key differences between langchain and semantic kernel?

**Comment by zs-dima** [comment]
@priya-27 
LangChain has many agents, tools, plugins etc. out of the box.
More over, LangChain has 10x more popularity, so has about 10x more developer activity to improve it.
On other hand, Semantic Kernel architecture and quality is better, that's quite promising for Semantic Kernel.
https://python.langchain.com/en/latest/modules/agents.html

**Reply by alvercau** [reply]
Could you elaborate a bit on why the architecture and quality of Semantic Kernel is better according to you? Thanks!

**Comment by samleegithub** [comment]
**BIG CAVEAT**: I am not part of the SK team and this is based on what I understand so far. Please feel free to correct. Looking forward to learn more about the SK roadmap. :)

I have found a few differences:

- Planner in SK is not the same as Agents in LangChain.
  - Agents in LangChain use recursive calls to the LLM to decide the next step to take based on the current state.
  - The two planner implementations in SK are not self-correcting.
    - Sequential planner tries to produce all the steps at the very beginning, so it is unable to handle unexpected errors.
    - Action planner only chooses one tool to satisfy the goal
  - I am hoping to see a new planner capability in SK that matches Action Agents in LangChain that recursively calls the LLM model until the goal is achieved.
    - https://python.langchain.com/docs/modules/agents/agent_types/
    - They use the framework described in the ReAct/MRKL paper: https://arxiv.org/pdf/2205.00445.pdf

- SQL
  - LangChain has a SQL agent to support conenctions to SQL databases.
  - SK doesn't have this yet.

- Math
  - LangChain implemented math calculations by using LLM to generate python code that is run in a python runtime.
  - SK doesn't have this yet

**Reply by germanbrunini** [reply]
Hi @SeaDude hope you are doing fine! This reply just spotted my concerns, as I have deployed a full architecture using langchain and now I am being asked to migrate to semantic-kernel. The thing is, 80% of my product relies on langchain agents. Specifically, a functionality that takes a dataframe as an input and, based on client "analytical queries", it creates an sql/python(pandas) agent that can address this query and output results in LLM format. Can you tell me if I am wrong to assume SK is not there yet for my app?

Thanks in advance!!

**Reply by samleegithub** [reply]
Looks like a stepwise planner is being worked on now.

https://github.com/microsoft/semantic-kernel/pull/1468

Also support for natural language to math calculations has been added.

https://github.com/microsoft/semantic-kernel/pull/1464

Now I'm just waiting on SQL agent support.

**Reply by joslat** [reply]
Hi, is this StepWise planner supports/implements the recursive behavior that @samleegithub  mentioned? Talking in regards to adding the recursive behavior from Agents in LangChain?  - the LangChain Agents (analogous to SK Planner) are doing recursive calls to the LLM to decide the next step based on the current step  outcome. This behavior is self-correcting, so it is able to handle unexpected errors. I find this would add a ton of value...

For example, if the first step did not produce a valid output and needs to be repeated... or this is detected at the third step and needs to repeat from step 2.

**Reply by samleegithub** [reply]
@joslat yes that is correct.

You can learn more here: https://youtube.com/watch?v=DG_Ge1v0c4Q&feature=share8

**Reply by leandroaldv** [reply]
https://learn.microsoft.com/en-us/semantic-kernel/overview/

**Reply by leandroaldv** [reply]
https://learn.microsoft.com/en-us/semantic-kernel/memories/vector-db?source=recommendations

**Comment by taoari** [comment]
@samleegithub I actually got some frustration on langchain agents. Langchain agents depend on a triple backtick output of JSON format for the next step to successfully execute. This fails so often even for GPT-3.5, not to say open-source LLMs (I would eventually hope to build my app based on open-source LLMs).  

To find a more robust alternative to Langchain agent. I found `guidance` and this repo. I really like guidance as it allows LLM to output in a fixed format, but guidance is not feature-rich, and you have to build all things from the ground up. Semantic Kernel seems to be a nice alternative to Langchain. I am new to this library. Is it more robust compared to langchain agents?

**Comment by formulahendry** [comment]
I am not part of the SK team, but I am also learning LangChain and SK recently.
I have created a repo (WIP) to compare Semantic Kernel and LangChain based on my current understanding:
https://github.com/formulahendry/semantic-kernel-vs-langchain
Hope it could help.

**Comment by anthonypuppo** [comment]
Used both pretty extensively the past few months so I'll throw my thoughts below in case it helps someone. Note that I may appear biased as I've almost entirely migrated off LangChain in favor of Semantic Kernel (don't think it's nearly customizable/discoverable enough).

Use LangChain when...
- Prototyping
- Comfortable with the defaults and lack of easy customization (though I'll say LangSmith is a HUGE step in the right direction)
- Your app is primarily based on natural language processing

Use Semantic Kernel when...
- You're comfortable writing code
- You need full control over the logic but could benefit from a thin boilerplate wrapper to help with coordinating all the pieces
- Existing, code only apps could benefit but natural language processing isn't the sole focus

**tl;dr** LangChain if trying to impress stakeholders with a demo, Semantic Kernel if you want to mix and match plain code with LLM's.

**Reply by deter3** [reply]
I did not see Langchain has any special expertise and depth knowledge in any fields they are covering .

**Comment by orchardc** [comment]
Another difference I guess is, that LangChain has a large community and you often get answers easily; SK is still new and requires a big community boost.

**Comment by WeihanLi** [comment]
There's a comparison found hours ago 
https://github.com/formulahendry/semantic-kernel-vs-langchain

**Comment by tonyqus** [comment]
Sorry, I have to ask this question: Can any MSFT show me if semantic-kernel can run on AWS? If this project is just to promote Azure instead of filling gap of LLM usage, I'm afraid it will definitely fail in the future.

**Reply by tonyqus** [reply]
And I reviewed the [implementation of semantic memory](https://github.com/microsoft/semantic-memory). All the backends it supports are Azure based. Do you have plan to support AWS S3 or AWS OpenSearch or any AWS service?

**Reply by tonyqus** [reply]
I opened a ticket in semantic memory repo https://github.com/microsoft/semantic-memory/issues/93

**Reply by anthonypuppo** [reply]
> Sorry, I have to ask this question: Can any MSFT show me if semantic-kernel can run on AWS? If this project is just to promote Azure instead of filling gap of LLM usage, I'm afraid it will definitely fail in the future.

Semantic Kernel is a .NET library, meaning it will run anywhere you can run .NET. On AWS you could use a multitude of different services to accomplish this (Lightsail, EC2, etc).

> And I reviewed the [implementation of semantic memory](https://github.com/microsoft/semantic-memory). All the backends it supports are Azure based. Do you have plan to support AWS S3 or AWS OpenSearch or any AWS service?

You can [create custom pipelines](https://github.com/microsoft/semantic-memory#custom-memory-ingestion-pipelines) to support things that SM doesn't include out of the box.
